### ISSUE TITLE
Publish client library to npm

### ISSUE THREAD

**Issue by webkonstantin** [umputun/remark42#240]
It would be nice to be able to use remark client library directly instead of dropping script tags on the page

Something like this

```js
const remark = require('remark')

remark.embed(node, options)
remark.counter(node, options)
remark.lastComments(node, options)
```

**Comment by igoradamenko** [comment]
Yep, we'll do it after switching from iframe widget to iframe-less (https://github.com/umputun/remark/issues/218#issuecomment-438005245). Right now it's overengineering to try to extract things from current widget to the package and to the iframe widget. 

I'm leaving this issue open as a reminder. Thanks.

**Comment by akellbl4** [comment]
Possible Requirements:

- pack code by widget with ability to lazy load the module out of the box
- provide all of the necessary methods for controlling the widgets
- pack file with API methods separately and expose API methods for outside use

Ideas:

- create wrappers for JS frameworks (for example remark42-react) for easier integration with them
- provide theme context which provides ability to modify theme of comments
- provide ability to replace components with user's own implementation

**Comment by paskal** [comment]
Closing this after seven years, as we have decided against publishing the widgets as an npm package.

The reason is authentication rather than packaging. The widget runs in an iframe served from the Remark42 origin, and the OAuth login flow depends on that: the popup authenticates against Remark42 first-party, and the cookie it sets is only usable by a widget running on that origin. Imported into a host page's bundle, the widget would run in the host's origin and its auth cookie would be third-party, which Safari blocks today and Chrome is moving away from. Login would break, and break silently.

The iframe-less direction mentioned in the first reply here did not happen, and the auth flow built since is the reason it will not.

What the request was actually after, control over when and how widgets load, is available today through `window.REMARK42.createInstance(config)` and the matching `destroy()`, which cover lazy loading, single-page app route changes and mounting on intersection. The contract is written out in #1714. For a fully custom frontend the supported path is the REST API.

Discussed further in #1715.
